### PR TITLE
Add possiblity for text output for sct-compute-snr

### LINKED ISSUE #3453: Add possibility for output TXT
**Feature justification**

Function `sct_compute_snr` outputs results in the Terminal:
~~~
julien-macbook:~/temp/20210712_gmchallenge/data_processed/9418 $ sct_compute_snr -i data_concat.nii.gz -method mult -m data1_crop_wmseg_erode.nii.gz 

--
Spinal Cord Toolbox (git-jca/sb/3063-csa-measure-from-pmj-05976283f121899e8903e8bb8222dce004940236)

sct_compute_snr -i data_concat.nii.gz -method mult -m data1_crop_wmseg_erode.nii.gz
--


SNR_mult = 20.003156441135573
~~~

When implementing pipelines, one needs to access this Terminal output and use it for making statistics. This currently requires an ugly regex manipulation to access the float value from stdout. 

**Describe the solution you'd like**

A new flag `-o` where user could indicate a TXT file name where results are output. The TXT file would only contain the number, so it can conveniently be opened by any Python script.

**Additional context**

The need for this feature came up while working on the [GM challenge project](https://github.com/sct-pipeline/gm-challenge).

### PR DESCRIPTION
**Pull request by santoshdvc**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [ ] I've applied a [release milestone](https://github.com/spinalcordtoolbox/spinalcordtoolbox/milestones) (major, minor, patch) in line with [Semantic Versioning guidelines](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Misc%3A-Creating-a-new-release#convention-for-naming-releases) 
- [x] I've assigned a reviewer

#### PR contents

- [x] I've consulted [SCT's internal developer documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki) to ensure my contribution is in line with any relevant design decisions
- [x] I've added [relevant tests](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Programming%3A-Tests) for my contribution
- [ ] I've updated the [relevant documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Programming%3A-Documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description

I have changed the `sct-compute-snr` file by adding an function for text file output. I have added the optional argument  `-o` with string argument. The output number of `snr_roi` is saved in TXT.
```
-o <file/path name>
```
I have changed the code to resolve the previous type error.
I created the branch by newly cloning the repository and starting fresh, I don't think this time any mode of file was changed.
Please verify it.

## Linked issues

Resolves #3453

### CODE REVIEW

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (jcohenadad) [inline spinalcordtoolbox/scripts/sct_compute_snr.py:94]:
```suggestion
    optional.add_argument(
        '-o',
        metavar=Metavar.str,
        type=str,
        default=None,
        help="File name where to write the computed SNR."
    )

```

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (jcohenadad) [inline spinalcordtoolbox/scripts/sct_compute_snr.py:135]:
```suggestion
    if arguments.o is not None:
        file_name = arguments.o
    else:
        file_name = None

```

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (jcohenadad) [inline spinalcordtoolbox/scripts/sct_compute_snr.py:115]:
```suggestion
```

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (jcohenadad) [inline spinalcordtoolbox/scripts/sct_compute_snr.py:204]: no need to add ext without user consent-- this is confusing for the user
```suggestion
```

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (jcohenadad) [inline spinalcordtoolbox/scripts/sct_compute_snr.py:205]:
```suggestion
        f = open(file_name,"w")
```

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_compute_snr.py:135]: If the default value for `arguments.o` is `None`, I think we can just do `file_name = arguments.o` and it will have the same effect.

Comment (jcohenadad) [reply spinalcordtoolbox/scripts/sct_compute_snr.py:135]: yes, i was not sure it would work (the problem with remote forks is that we cannot commit the changes and immediately test them)-- but yes, let's do it that way-- thank you

Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_compute_snr.py:135]: > (the problem with remote forks is that we cannot commit the changes and immediately test them)

That's fair! Yeah, there are only a couple of ways to do this:

* Create a local branch that points to the fork's branch, then make changes locally and test them locally.
* Or, if @santoshdvc has checked the box [**"Allow edits from maintainers"**](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests)), then we can make changes directly to the fork's branch.

Comment (jcohenadad) [reply spinalcordtoolbox/scripts/sct_compute_snr.py:135]: > Or, if @santoshdvc has checked the box "Allow edits from maintainers"), then we can make changes directly to the fork's branch.

oh! great feature! @santoshdvc could you please do this?

**Comment by santoshdvc** [comment]
I did comitted the suggested changes. Most of the check passed but I can't get the Python Linting check successful. It fails at a point when it approaches certain file.
```
./contrib/fsl_integration/sct_plugin.py:789:1: E305 expected 2 blank lines after class or function definition, found 1

Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.8.10/x64/bin/lintly", line 8, in <module>
    sys.exit(main())
  File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/lintly/cli.py", line 74, in main
    build.execute()
  File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/lintly/builds.py", line 82, in execute
    self.submit_to_pr(patch)
  File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/lintly/builds.py", line 145, in submit_to_pr
    self.git_client.create_check_run(
  File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/lintly/backends/github.py", line 249, in create_check_run
    response = client.post(url, data, headers={'Accept': GITHUB_CHECKS_HEADER})
  File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/lintly/backends/github.py", line 81, in post
    return self._do_request('post', url, json.dumps(data), headers)
  File "/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/lintly/backends/github.py", line 113, in _do_request
    raise GitClientError(response.content, status_code=response.status_code)
lintly.backends.errors.GitClientError: b'{"message":"Resource not accessible by integration","documentation_url":"https://docs.github.com/rest/reference/checks#create-a-check-run"}'
Error: Process completed with exit code 1.
```

**Comment by joshuacwnewton** [comment]
> I did comitted the suggested changes. Most of the check passed but I can't get the Python Linting check successful. It fails at a point when it approaches certain file.

My apologies, this is a problem on our end, see: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3462.

Feel free to ignore the error. :slightly_smiling_face:

**Comment by santoshdvc** [comment]
Okay. I have removed this request as draft and added it for review. I think once the problem is solved, this can be merged. Am I right?

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (joshuacwnewton) [inline spinalcordtoolbox/scripts/sct_compute_snr.py:135]: Nitpick, but this is a simpler way of writing this section:

```suggestion
    file_name = arguments.o
```

Because the `None` code is redundant (`if arguments.o is None`, then the result will be the same).

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (joshuacwnewton) [inline spinalcordtoolbox/scripts/sct_compute_snr.py:92]: Just a small grammar nitpick...

```suggestion
        help="File name to write the computed SNR to."
```

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (joshuacwnewton) [inline spinalcordtoolbox/scripts/sct_compute_snr.py:108]: Nitpick: Could you please add this empty line back?

```diff
+                                                                                                  
```

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (joshuacwnewton) [inline spinalcordtoolbox/scripts/sct_compute_snr.py:207]: This pattern can be simplified by using Python's [`with` statement](https://realpython.com/python-with-statement/):

```suggestion
        with open(file_name, "w") as f:
            f.write(str(snr_roi))
            printv('\nFile saved to ' + file_name)
```

That way, you don't need to write `f.close`.

File: spinalcordtoolbox/scripts/sct_compute_snr.py
Comment (joshuacwnewton) [inline spinalcordtoolbox/scripts/sct_compute_snr.py:209]: One other suggestion: Could you please update the existing test? I've linked to the existing test file below:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/24b6ca2511e9a4163da1fdb7565f9a39bbf1d267/testing/cli/test_cli_sct_compute_snr.py#L11-L15

We can add the `-o` argument to this test, then check to make sure the computed SNR matches an expected value. Here is a rough idea of what the updated test could look like:

```python
 def test_sct_compute_snr_against_groundtruth(): 
     """Run the CLI script and check SNR against a ground truth value.""" 
     fname_out = "computed_snr.txt"
     sct_compute_snr.main(argv=['-i', 'dmri/dwi.nii.gz', '-m', 'dmri/dmri_T0001.nii.gz', '-method', 'diff', 
                                '-vol', '0,5', '-o', fname_out])
    with open(fname_out, "r") as f:
         snr = f.read()
    assert snr == pytest.approx(1.234) # Replace with whatever the SNR is supposed to be
``` 

We also have the [Testing](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Testing) page in our developer wiki which provides some tips for writing new tests. (But, this is a simple test, so you probably won't need to do anything too complicated.)

**Review by joshuacwnewton — CHANGES_REQUESTED** [review]
> Okay. I have removed this request as draft and added it for review. I think once the problem is solved, this can be merged. Am I right?

Yes, exactly. Once we review your PR, and all the feedback is addressed, we will mark it as "Approved", then it can be merged into the project. :slightly_smiling_face:

**Comment by santoshdvc** [comment]
Hi, 
I have added the test function and resolved the changes asked. 
Also, I tried the spinalcordtoolbox myself giving them the input data used for tests. I am attaching the screenshots of results I got.
compute_output

**Comment by jcohenadad** [comment]
> I have added the test function and resolved the changes asked.

@santoshdvc I don't see the test file updated?

**Comment by jcohenadad** [comment]
@joshuacwnewton i'm trying again your suggestion, this time without the typo (hopefully), but still running into problems:
~~~
julien-macbook:~/code/sct $ gb
  jca/sb/3063-csa-measure-from-pmj
  jn/3426-update-sct_plugin.py-fsleyes-v1.0
* master
  release
  sb/3063-csa-measure-from-pmj
julien-macbook:~/code/sct $ git pull
Already up to date.
julien-macbook:~/code/sct $ git remote add santoshdvc git@github.com:santoshdvc/spinalcordtoolbox.git
fatal: remote santoshdvc already exists.
julien-macbook:~/code/sct $ git fetch santoshdvc
fatal: 'git@github.santoshdvc/spinalcordtoolbox.git' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
julien-macbook:~/code/sct $ git checkout -b santoshdvc-add-text-output-2 santoshdvc/add-text-output-2
fatal: 'santoshdvc/add-text-output-2' is not a commit and a branch 'santoshdvc-add-text-output-2' cannot be created from it
~~~

out of curiosity: why do you recommend not using the lines suggested by github (which work for me)?

**Review by jcohenadad — APPROVED** [review]
LGTM! thank you for your contribution @santoshdvc 🏅

**Comment by joshuacwnewton** [comment]
> i'm trying again your suggestion, this time without the typo (hopefully), but still running into problems:

Reading your error messages, it looks like the old remote (with the typo) still exists, and it's preventing you from creating a new remote:

```console
julien-macbook:~/code/sct $ git remote add santoshdvc git@github.com:santoshdvc/spinalcordtoolbox.git
fatal: remote santoshdvc already exists.
```

So, what you want to do is remove the existing remote first, then add the non-typo remote:

```console
> git remote rm santoshdvc
> git remote add santoshdvc git@github.com:santoshdvc/spinalcordtoolbox.git
> git fetch santoshdvc
remote: Enumerating objects: 124, done.
remote: Counting objects: 100% (124/124), done.
remote: Compressing objects: 100% (74/74), done.
remote: Total 124 (delta 72), reused 81 (delta 48), pack-reused 0
Receiving objects: 100% (124/124), 26.83 KiB | 915.00 KiB/s, done.
Resolving deltas: 100% (72/72), completed with 15 local objects.
From github.com:santoshdvc/spinalcordtoolbox
[...]
> git checkout -b add-text-output-2_local santoshdvc/add-text-output-2
Branch 'add-text-output-2_local' set up to track remote branch 'add-text-output-2' from 'santoshdvc'.
Switched to a new branch 'add-text-output-2_local'
```

**Comment by joshuacwnewton** [comment]
> out of curiosity: why do you recommend not using the lines suggested by github (which work for me):

When you create a new local branch like this: 

```console
julien-macbook:~/code/sct $ git checkout -b santoshdvc-add-text-output-2 master
```

This local branch will point at whichever commit you were previously checked out at.

Then, when you pull the branch from the fork repo:

```console
julien-macbook:~/code/sct $ git pull https://github.com/santoshdvc/spinalcordtoolbox.git add-text-output-2
```

You're asking git to `git fetch` and `git merge` the changes from the fork's branch into your local branch. 

I can think of 2 downsides to this:

1. `git merge` could go wrong. There could be conflicts, an unnecessary merge commit, etc. 
   * (It would depend on where exactly you created your local branch from. But, the point is, as long as `git merge` is involved, there is some possibility of error.)
   * My suggested method instead creates the local branch _directly_ from the fork's branch, without the need to `git merge`.
2. The local branch won't be set up to track the fork's branch.
   * My suggested method instead creates a [remote-tracking branch](https://git-scm.com/book/en/v2/Git-Branching-Remote-Branches):

       > Remote-tracking branches are references to the state of remote branches. They’re local references that you can’t move; Git moves them for you whenever you do any network communication, to make sure they accurately represent the state of the remote repository. Think of them as bookmarks, to remind you where the branches in your remote repositories were the last time you connected to them. 
    * You can tell that a remote-tracking branch was made because of the text `"Branch 'add-text-output-2_local' set up to track remote branch 'add-text-output-2' from 'santoshdvc'."` in the console log from https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3455#issuecomment-881595200.

**Comment by joshuacwnewton** [comment]
Thank you so much for the contribution, @santoshdvc! :heart:

**Comment by joshuacwnewton** [comment]
> i'm trying again your suggestion, this time without the typo (hopefully), but still running into problems:

Reading your error messages, it looks like the old remote (with the typo) still exists, and it's preventing you from creating a new remote:

```console
julien-macbook:~/code/sct $ git remote add santoshdvc git@github.com:santoshdvc/spinalcordtoolbox.git
fatal: remote santoshdvc already exists.
```

So, what you want to do is remove the existing remote first, then add the non-typo remote:

```console
> git remote rm santoshdvc
> git remote add santoshdvc git@github.com:santoshdvc/spinalcordtoolbox.git
> git fetch santoshdvc
remote: Enumerating objects: 124, done.
remote: Counting objects: 100% (124/124), done.
remote: Compressing objects: 100% (74/74), done.
remote: Total 124 (delta 72), reused 81 (delta 48), pack-reused 0
Receiving objects: 100% (124/124), 26.83 KiB | 915.00 KiB/s, done.
Resolving deltas: 100% (72/72), completed with 15 local objects.
From github.com:santoshdvc/spinalcordtoolbox
[...]
> git checkout -b add-text-output-2 santoshdvc/add-text-output-2
Branch 'add-text-output-2' set up to track remote branch 'add-text-output-2' from 'santoshdvc'.
Switched to a new branch 'add-text-output-2'
```

**Comment by joshuacwnewton** [comment]
> out of curiosity: why do you recommend not using the lines suggested by github (which work for me):

When you create a new local branch like this: 

```console
julien-macbook:~/code/sct $ git checkout -b santoshdvc-add-text-output-2 master
```

This local branch will point at whichever commit you were previously checked out at.

Then, when you pull the branch from the fork repo:

```console
julien-macbook:~/code/sct $ git pull https://github.com/santoshdvc/spinalcordtoolbox.git add-text-output-2
```

You're asking git to `git fetch` and `git merge` the changes from the fork's branch into your local branch. 

But, `git merge` could go wrong. There could be conflicts, an unnecessary merge commit, etc. (It would depend on where exactly you created your local branch from, and how that relates to the fork's branch. But, the point is, as long as `git merge` is involved, there is some possibility of error.)

My suggested method instead creates the local branch _directly_ from the fork's branch, without the need for the intermediate `git merge` step, so you can trust that your local branch will be identical to the fork's branch.

**Comment by santoshdvc** [comment]
Thanks @joshuacwnewton @jcohenadad . I learned lot from you.

**Comment by jcohenadad** [comment]
and thank you for [your explanation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3455#issuecomment-881600303) @joshuacwnewton ! i also learned a lot 😊

**Comment by joshuacwnewton** [comment]
@jcohenadad I just wanted to mention 1 other benefit to adding a contributor's fork as a `remote`: 

Once you have the remote set up, it makes it easy to checkout other branches, too, if they contribute multiple times. For example, all I needed to do to test out https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3473 was:

```console
> git fetch santoshdvc
remote: Enumerating objects: 17, done.
remote: Counting objects: 100% (17/17), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 10 (delta 7), reused 6 (delta 4), pack-reused 0
Unpacking objects: 100% (10/10), 1.29 KiB | 331.00 KiB/s, done.
From github.com:santoshdvc/spinalcordtoolbox
 * [new branch]        af/add_ref_harrison               -> santoshdvc/af/add_ref_harrison
 * [new branch]        af/add_ref_vallotton              -> santoshdvc/af/add_ref_vallotton
   d76783ce..452a3707  ah/docs-aesthetic                 -> santoshdvc/ah/docs-aesthetic
   9eb1c876..26106c65  jca/sb/3063-csa-measure-from-pmj  -> santoshdvc/jca/sb/3063-csa-measure-from-pmj
 * [new branch]        jn/2702-fix-code-coverage         -> santoshdvc/jn/2702-fix-code-coverage
 * [new branch]        jn/demo-upstream-ivadomed-torch-changes -> santoshdvc/jn/demo-upstream-ivadomed-torch-changes
 * [new branch]        jn/repo-clean-up                  -> santoshdvc/jn/repo-clean-up
 + 5bbff6a3...3deecead master                            -> santoshdvc/master  (forced update)
   27a9b242..c1c4ee9c  sb/3063-csa-measure-from-pmj      -> santoshdvc/sb/3063-csa-measure-from-pmj
 * [new branch]        sb/3456-improve-R-L-placement-pmj -> santoshdvc/sb/3456-improve-R-L-placement-pmj
 * [new branch]        sb/add-qc-for-pmj-csa             -> santoshdvc/sb/add-qc-for-pmj-csa
 * [new branch]        scary-message-removal             -> santoshdvc/scary-message-removal
> git checkout scary-message-removal
Branch 'scary-message-removal' set up to track remote branch 'scary-message-removal' from 'santoshdvc'.
Switched to a new branch 'scary-message-removal'

**Pull request merged**
